File: tutorials/vbnet/loading_balance/lb_worker/piworker.vb
Imports System
Imports SocketProAdapter
Imports SocketProAdapter.ClientSide

Public Class PiWorker
    Inherits CAsyncServiceHandler

    Public Sub New()
        MyBase.New(piConst.sidPiWorker)
    End Sub

    Protected Overrides Sub OnResultReturned(ByVal sRequestId As UShort, ByVal UQueue As CUQueue)
        If RouteeRequest Then
            Select Case sRequestId
                Case piConst.idComputePi
                    Dim dStart As Double
                    Dim dStep As Double
                    Dim nNum As Integer
                    UQueue.Load(dStart).Load(dStep).Load(nNum)
                    Dim dX As Double = dStart + dStep / 2
                    Dim dd As Double = dStep * 4.0
                    Dim ComputeRtn As Double = 0.0
                    For n As Integer = 0 To nNum - 1
                        dX += dStep
                        ComputeRtn += dd / (1 + dX * dX)
                    Next n
                    SendRouteeResult(ComputeRtn)
                Case Else
            End Select
        End If
    End Sub
End Class

File: tutorials/java/src/loading_balance/worker/PiWorker.java
package loading_balance.worker;

import loading_balance.piConst;
import SPA.*;
import SPA.ClientSide.CAsyncServiceHandler;

public class PiWorker extends CAsyncServiceHandler {

    public PiWorker() {
        super(loading_balance.piConst.sidPiWorker);
    }

    @Override
    protected void OnResultReturned(short sRequestId, CUQueue UQueue) {
        if (getRouteeRequest()) {
            switch (sRequestId) {
                case piConst.idComputePi: {
                    double dStart = UQueue.LoadDouble();
                    double dStep = UQueue.LoadDouble();
                    int nNum = UQueue.LoadInt();
                    double dX = dStart + dStep / 2;
                    double dd = dStep * 4.0;
                    double ComputeRtn = 0.0;
                    for (int n = 0; n < nNum; n++) {
                        dX += dStep;
                        ComputeRtn += dd / (1 + dX * dX);
                    }
                    try (CScopeUQueue sb = new CScopeUQueue()) {
                        SendRouteeResult(sb.Save(ComputeRtn).Save(dStart));
                    }
                }
                break;
                default:
                    break;
            }
        }
    }
}
